### PR TITLE
Improving usability on onboarding screen when using keyboard and screen reader

### PR DESCRIPTION
**Pull request by Rafi993**
## What type of PR is this?

- [ ] Refactor
- [ ] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
Improving usability when using keyboard and screen reader

## Related Tickets & Documents

Closes https://github.com/forem/forem/issues/9585

## QA Instructions, Screenshots, Recordings

1. Go to url http://localhost:3000/onboarding
2. Try to complete the on-boarding journey with keyboard (with screen reader on)

## Added tests?

- [x] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [ ] docs.forem.com
- [ ] readme
- [x] no documentation needed

## What gif best describes this PR or how it makes you feel?

Happy

### CODE REVIEW

File: app/assets/stylesheets/preact/onboarding-modal.scss [hunk 1]
```diff
@@ -246,7 +246,7 @@
 }
 // modal overflow scrolling
 .onboarding-modal-scroll-container {
-  height: calc(100vh - 289px);
+  height: calc(100vh - 332px);
```
Comment (Rafi993) [inline app/assets/stylesheets/preact/onboarding-modal.scss:249]: Since it was causing two scrollbars I changed it to so that it has only one scrollbar

File: app/javascript/onboarding/components/FollowTags.jsx [hunk 2]
```diff
@@ -118,9 +118,9 @@ class FollowTags extends Component {
               <h2 className="subtitle">Follow tags to customize your feed</h2>
             </header>
             <div className="onboarding-modal-scroll-container">
-              <div data-testid="onboarding-tags" className="onboarding-tags">
+              <ul data-testid="onboarding-tags" className="onboarding-tags">
```
Comment (Rafi993) [inline app/javascript/onboarding/components/FollowTags.jsx:121]: I changed div to ul and li so that I could be read by screen reader

File: app/javascript/shared/components/focustrap.jsx [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import PropTypes from 'prop-types';
+import { h, Fragment } from 'preact';
+import { useEffect, useRef } from 'preact/hooks';
+import { createFocusTrap } from 'focus-trap';
+import { defaultChildrenPropTypes } from '../../common-prop-types';
+
+const FocusTrap = ({ selector, children }) => {
```
Comment (Rafi993) [inline app/javascript/shared/components/focustrap.jsx:7]: I created this wrapper component that uses library 'focus-trap' to trap focus. Ideally I love to  convert it to a hook rather than a component but since it is used by class components too it can't be a hook.

File: app/javascript/shared/components/focustrap.jsx [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import PropTypes from 'prop-types';
+import { h, Fragment } from 'preact';
+import { useEffect, useRef } from 'preact/hooks';
+import { createFocusTrap } from 'focus-trap';
```
Comment (Rafi993) [inline app/javascript/shared/components/focustrap.jsx:4]: I added this library https://bundlephobia.com/result?p=focus-trap@6.1.2  It's size is **2.3kb** minified and gzipped is this ok?

File: app/javascript/shared/components/focustrap.jsx [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    return () => {
+      focusTrap.current.deactivate();
+    };
+  });
+
+  return <Fragment>{children}</Fragment>;
+};
+
+FocusTrap.defaultProps = {
+  selector: '.crayons-modal',
```
Comment (Rafi993) [inline app/javascript/shared/components/focustrap.jsx:22]: I made the default selector value `'.crayon-modal'` since this component will be mostly be used by modals (rarely in other places where different selector string could be passed)

File: app/javascript/onboarding/components/FollowTags.jsx (hunk 2 above)
Comment (Rafi993) [inline app/javascript/onboarding/components/FollowTags.jsx]: I changed div to ul and li so that I could be read by screen reader

File: app/javascript/shared/components/focusTrap.jsx [hunk 6]
```diff
@@ -0,0 +1,30 @@
+import PropTypes from 'prop-types';
+import { h, Fragment } from 'preact';
+import { useEffect, useRef } from 'preact/hooks';
+import { createFocusTrap } from 'focus-trap';
+import { defaultChildrenPropTypes } from '../../common-prop-types';
+
+const FocusTrap = ({ selector, children }) => {
```
Comment (Rafi993) [inline app/javascript/shared/components/focusTrap.jsx:7]: I created this wrapper component that uses library 'focus-trap' to trap focus. Ideally I love to convert it to a hook rather than a component but since it is used by class components too it can't be a hook.

File: app/javascript/shared/components/focusTrap.jsx [hunk 7]
```diff
@@ -0,0 +1,30 @@
+import PropTypes from 'prop-types';
+import { h, Fragment } from 'preact';
+import { useEffect, useRef } from 'preact/hooks';
+import { createFocusTrap } from 'focus-trap';
```
Comment (Rafi993) [inline app/javascript/shared/components/focusTrap.jsx:4]: I added this library https://bundlephobia.com/result?p=focus-trap@6.1.2 It's size is 2.3kb minified and gzipped is this ok?

File: app/javascript/shared/components/focusTrap.jsx [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    return () => {
+      focusTrap.current.deactivate();
+    };
+  });
+
+  return <Fragment>{children}</Fragment>;
+};
+
+FocusTrap.defaultProps = {
+  selector: '.crayons-modal',
```
Comment (Rafi993) [inline app/javascript/shared/components/focusTrap.jsx:22]: I made the default selector value '.crayon-modal' since this component will be mostly be used by modals (rarely in other places where different selector string could be passed)

File: app/javascript/onboarding/components/FollowTags.jsx [hunk 9]
```diff
@@ -100,84 +101,97 @@ class FollowTags extends Component {
...
+              <header className="onboarding-content-header">
+                <h1 id="title" className="title">
+                  What are you interested in?
+                </h1>
+                <h2 id="subtitle" className="subtitle">
+                  Follow tags to customize your feed
+                </h2>
+              </header>
+              <div className="onboarding-modal-scroll-container">
+                <ul data-testid="onboarding-tags" className="onboarding-tags">
```
Comment (Rafi993) [inline app/javascript/onboarding/components/FollowTags.jsx:132]: I changed div to ul and li so that I could be read by screen reader

**Review by rhymes — COMMENTED** [review]
Haven't reviewed the code but I wonder if we can improve the behavior in the third slide

I have to tab 3 times before I can press space to select the "follow" button:

on1

It seems to work nice otherwise. There are a few violations found in the Accessibility panel in Firefox but I think they are beyond the objective of this PR.

Thanks @Rafi993!

File: app/javascript/shared/components/focusTrap.jsx (hunk 8 above)
Comment (Link2Twenty) [reply app/javascript/shared/components/focusTrap.jsx:22]: If you open a second modal over the top of an existing one I think the tabbing will be locked to the first.
That being said I believe, with how you've set it up, you can some something like

```jsx
<FocusTrap selector={reference.current}>
  <div id="locked-in" ref={reference}></div>
</FocusTrap>
```

**Comment by Link2Twenty** [comment]
FocusTrap would be useful for #10610 too.

File: app/javascript/shared/components/focusTrap.jsx (hunk 6 above)
Comment (Link2Twenty) [inline app/javascript/shared/components/focusTrap.jsx:7]: Is it worth having a `bool` active prop?
Meaning in the future we could add the ability to lock focus if we want to,

Comment (Rafi993) [reply app/javascript/shared/components/focusTrap.jsx:7]: I think boolean `active` can be added to **FocusTrap** component and made optional default value being `true`. But right now it is not used here but it could be added later to the component if needed without affecting existing code.

File: app/javascript/shared/components/focusTrap.jsx (hunk 8 above)
Comment (Rafi993) [reply app/javascript/shared/components/focusTrap.jsx:22]: Ref would work too. Incase modal appears another modal it can pass another selector string like id of the modal and it would work.

```
<FocusTrap selector="#secondModal">
  <div>
    second modal
  </div>
</FocusTrap>
```
But on the other hand nested modals might not be used anywhere (but I could be wrong). Having no prop needed to be passed for this component by default makes API much simpler.

**Comment by Rafi993** [comment]
> Haven't reviewed the code but I wonder if we can improve the behavior in the third slide
> 
> I have to tab 3 times before I can press space to select the "follow" button:
> 
> on1
> 
> It seems to work nice otherwise. There are a few violations found in the Accessibility panel in Firefox but I think they are beyond the objective of this PR.
> 
> Thanks @Rafi993!

When "follow" button is selected focus is jumping to the previous element causing this issue. Let me search for reason it is happening.

**Comment by Rafi993** [comment]
> Haven't reviewed the code but I wonder if we can improve the behavior in the third slide
> 
> I have to tab 3 times before I can press space to select the "follow" button:
> 
> on1
> 
> It seems to work nice otherwise. There are a few violations found in the Accessibility panel in Firefox but I think they are beyond the objective of this PR.
> 
> Thanks @Rafi993!

Similar issue is also happening in tag selection too. I couldn't figure out why it is happening. Any thoughts @forem/oss

tags

Thanks for you help

File: app/javascript/onboarding/components/FollowUsers.jsx [hunk 10]
```diff
@@ -140,72 +141,86 @@ class FollowUsers extends Component {
...
+                </h2>
+              </header>
+
+              <ul
+                data-testid="onboarding-users"
+                className="onboarding-modal-scroll-container"
+              >
+                {users.map((user) => (
+                  <li key={user.name}>
+                    <button
```
Comment (Link2Twenty) [inline app/javascript/onboarding/components/FollowUsers.jsx:178]: I think this button should have tabIndex="-1".
I think we only want the follow button to be in the tab order.

Comment (Link2Twenty) [reply app/javascript/onboarding/components/FollowUsers.jsx:178]: Here is the flow with the tabindex -1

forem__outer-button

only 1 tab to get to the next follow button

Comment (Rafi993) [reply app/javascript/onboarding/components/FollowUsers.jsx:178]: Thank you @Link2Twenty. But there is one issue though the screen reader is not reading out items in the list so the user (the name of the user and the description below now) it just says follow. This might cause confusion for users who are using screen readers since they don't know which user did they just followed. I can add **aria-label** to the follow button saying 
**"Follow Jere Thompson"** it will help but it will still miss out the description but I don't know if that is ok.

Comment (Rafi993) [reply app/javascript/onboarding/components/FollowUsers.jsx:178]: When I remove the handler ```this.handleClick``` it is not happening I did try adding both ```event.preventDefault()``` and ```event.stopPropagation()``` they don't seem to affect this. It is something to do with the state change. Let me keep searching.

Comment (Rafi993) [reply app/javascript/onboarding/components/FollowUsers.jsx:178]: It is a bug in the ```FocusTrap``` component that is causing this when I remove it is working properly. Let me keep searching further.

Comment (Rafi993) [reply app/javascript/onboarding/components/FollowUsers.jsx:178]: I was able to fix this by moving FocusTrap component one level up to **app/javascript/onboarding/Onboarding.jsx** there are some tests failing (even though what the test is looking for is working when tested manually) let me fix those too.

Comment (Rafi993) [reply app/javascript/onboarding/components/FollowUsers.jsx:178]: I have fixed the breaking tests too

**Comment by Rafi993** [comment]
> Haven't reviewed the code but I wonder if we can improve the behavior in the third slide
> 
> I have to tab 3 times before I can press space to select the "follow" button:
> 
> on1
> 
> It seems to work nice otherwise. There are a few violations found in the Accessibility panel in Firefox but I think they are beyond the objective of this PR.
> 
> Thanks @Rafi993!

I have resolved this now

tab_key

**Comment by Rafi993** [comment]
I think some unrelated test failed.

**Comment by rhymes** [comment]
@Rafi993 the tests are fine, it's Codeclimate complaining :) We should likely ignore it in the context of this PR

**Review by rhymes — APPROVED** [review]
It looks good, thanks @Rafi993, only a question still on the "follow people slide". Is it possible to go down from "2 tabs + space to select follow" to "1 tab plus space"?

**Comment by Rafi993** [comment]
> It looks good, thanks @Rafi993, only a question still on the "follow people slide". Is it possible to go down from "2 tabs + space to select follow" to "1 tab plus space"?

Solution @Link2Twenty suggested above does it https://github.com/forem/forem/pull/10661#discussion_r501755695 . But when I was using screen reader it just read out the text in the "Follow" button alone and it did not read the user name or the description below.

File: app/javascript/onboarding/components/FollowUsers.jsx [hunk 11]
```diff
@@ -140,72 +141,86 @@ class FollowUsers extends Component {
...
+                        <h4 className="user-name">{user.name}</h4>
+                        <p className="user-summary">
+                          {he.unescape(user.summary || '')}
+                        </p>
+                      </div>
+                      <button
+                        data-testid="onboarding-user-following-status"
+                        type="button"
+                        className="user-following-status"
+                        aria-pressed={selectedUsers.includes(user)}
```
Comment (Link2Twenty) [inline app/javascript/onboarding/components/FollowUsers.jsx]: Here would could add 

```jsx
aria-label={`${selectedUsers.includes(user) ? 'Unfollow' : 'Follow'} ${user.name}`}
```

This will tell the screen reader what to read.

Comment (Link2Twenty) [reply app/javascript/onboarding/components/FollowUsers.jsx]: This would be to coincide with the `tabIndex` change.

Comment (Rafi993) [reply app/javascript/onboarding/components/FollowUsers.jsx]: > Here would could add
> 
> ```js-jsx
> aria-label={`${selectedUsers.includes(user) ? 'Unfollow' : 'Follow'} ${user.name}`}
> ```
> 
> This will tell the screen reader what to read.

 But it would miss the description for each user

Comment (Link2Twenty) [reply app/javascript/onboarding/components/FollowUsers.jsx]: I guess the three options are

- We remove something from the tabIndex be it the inner or outer button
- We accept multiple tabs is a thing that people have to deal with
- We wait for #10468 and use the useListNavigation hook to add j and k navigation support.

**Comment by nickytonline** [comment]
I'm looking to give this a review this week. Hacktoberfest was pretty busy. 😅

File: app/javascript/onboarding/components/FollowUsers.jsx (hunk 10 above)
Comment (marcysutton) [reply app/javascript/onboarding/components/FollowUsers.jsx:178]: To answer your comment regarding the button name, @Rafi993, an `aria-label` would help to uniquely identify the individual follow buttons. That would solve an issue that exists throughout the site where controls are repeated (i.e. "Links with the same name have a similar purpose" in the axe extension).

Comment (Rafi993) [reply app/javascript/onboarding/components/FollowUsers.jsx:178]: Thank you @marcysutton I have made the change. I did not know about this before.

File: package.json [hunk 12]
```diff
@@ -155,7 +155,7 @@
     "@honeybadger-io/webpack": "^1.2.0",
     "@rails/webpacker": "5.2.1",
     "@testing-library/user-event": "^12.1.7",
-    "ahoy.js": "^0.3.7",
+    "ahoy.js": "^0.3.8",
```
Comment (nickytonline) [inline package.json:158]: refactor (blocking): This looks like a bad merge or maybe ensure you have merged latest as this does not appear to be related to the work being done in this PR.

File: yarn.lock [hunk 13]
```diff
@@ -18039,10 +18039,10 @@ tabbable@^5.1.1:
   resolved "https://registry.yarnpkg.com/tabbable/-/tabbable-5.1.1.tgz#33d6950a615bc78c6eb6828f3953e66544817dc4"
   integrity sha512-qjkrV7I29Lkg///jPQnq26+26q9wQOKTEcckrwPejTsxxl2zdK60x0XmiEELCEParLhgbVQRflyKkpr+K/SmCQ==
 
-table@^5.2.3:
-  version "5.4.6"
-  resolved "https://registry.yarnpkg.com/table/-/table-5.4.6.tgz#1292d19500ce3f86053b05f0e8e7e4a3bb21079e"
-  integrity sha512-wmEc8m4fjnob4gt5riFRtTu/6+4rSe12TpAELNSqHMfF3IqnA+CH37USM6/YR3qRZv7e56kAEAtd6nKZaxe0Ug==
+table@^6.0.4:
```
Comment (nickytonline) [inline yarn.lock:18042]: refactor (blocking): This looks like a bad merge or maybe ensure you have merged latest as this does not appear to be related to the work being done in this PR.

**Review by nickytonline — CHANGES_REQUESTED** [review]

**Comment by Rafi993** [comment]
As we discussed on the stream it is better to split this PR into multiple small PR's so that it will be easy to review. I'll start creating separate small PR's.

**Comment by Rafi993** [comment]
Closing this PR in favour of smaller PRs that have been raised.
